### PR TITLE
Multi selection

### PR DESCRIPTION
**Pull request by Bdtrotte**
Some small general improvements for all tools: Cell signals are now sent when clicking outside the map, allowing one to draw a shape that extends outside the map (as expected). Also, all cell signals now include the event object from the original Widget event.

Selection vastly improved!
-Can now select multiple tiles at a tile
-Intuitive drag behavior
-Modifiers supported
-Multi clicking supported

Explore the commit messages for more details.

### CODE REVIEW

File: WallsAndHoles/tilemaphelpers.h [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+#include "tilemap.h"
+
+#include <QStack>
+#include <QSet>
+
+// QPoints are not hashable in Qt by default!
+inline uint qHash (const QPoint & key)
+{
+    return qHash (QPair<int,int>(key.x(), key.y()) );
+}
```
Comment (timoffex) [inline WallsAndHoles/tilemaphelpers.h:10]: I actually added this in array2dtools.h in my branch! Maybe we should put this into a separate header, and include it in both this header and my array2dtools.h header.

File: WallsAndHoles/tilemaphelpers.h [hunk 2]
```diff
@@ -0,0 +1,37 @@
+#ifndef TILEMAPHELPERS_H
+#define TILEMAPHELPERS_H
+
+#include "tilemap.h"
+
+#include <QStack>
+#include <QSet>
```
Comment (timoffex) [inline WallsAndHoles/tilemaphelpers.h]: Since you're not using QSets and QStacks in this header, I would move these includes into the cpp file.

**Review by timoffex — APPROVED** [review]
Everything looks good, just review my two comments. Merge is okay if there are no conflicts.

File: WallsAndHoles/tilemaphelpers.h (hunk 1 above)
Comment (Bdtrotte) [reply WallsAndHoles/tilemaphelpers.h:10]: Not a bad idea. It's so small though I'm not to worried if it were to just exist in both places, considering it is inline.

We can worry about it when merging your branch

**Pull request merged**